### ISSUE TITLE
libgraal 20.3.1 seems to fail in "compiling" in due time on window

### ISSUE THREAD

**Issue by spangaer** [oracle/graal#3142]
**Describe the issue**
I have some logging initializer code. There shouldn't be much special about it. It is loaded with a classloader by some jar sitting on the ext classpath (so between boot and regular classpath).

When running this code without flags (both Graal 19.3.4 & Graal 20.3.1) after 30 seconds something (Akka SFL4J logging) times out, and only then for some reason things start moving.
If you specify `-XX:-UseJVMCINativeLibrary` it starts in about 2 seconds.

**Steps to reproduce the issue**
I don't really know how to offer you a reproducer at the moment. As I wouldn't know where it goes wrong exactly.
I know there's some code in a class initializer kicks a thread in to motion to call a method on itself, which otherwise ends-up in a sort of recursion deadlock which might be tricky to handle for compilers. It actually looks like a self-deadlock of some sort here too.
Another remark is that all common logging frameworks are overridden or configured (JUL) to be diverted to SLF4J.

**Describe GraalVM and your environment:**
 - GraalVM version : **CE 19.3.4 CE 20.3.1**
 - JDK major version: **8**
 - OS: **Win 10 2004**
 - Architecture: **AMD64**

**More details**
If you want me to collect more data, I welcome instructions. I haven't bothered testing it on Linux, but if that can generate additional insights, that could be done. I have a workaround, which happens to be our default anyway, but I thought someone would like to know about this bug(?).

**Comment by spangaer** [comment]
A bit sharing
```scala
  implicit def system: ActorSystem = _system

  protected def init() {
    // enable process monitoring accessible from JMX
    ProcessStatsMonitorAkkaBean()
  }

  {
    // init actor system, needs to be late binding as the system needs slf4j itself
    val t = new Thread() {
      override def run() {
        // when libgraal is active this seems to run in to a self deadlock
        if (AgentConfig.isLibGraal)
          Thread.sleep(1.seconds.toMillis)

        init()
      }
    }
    t.setDaemon(true)
    t.start()
  }

...

  /** check if libgraal is active */
  lazy val isLibGraal = {
    ("graal" == System.getProperty("jvmci.Compiler")) && {
      val toCheck = ManagementFactory.getRuntimeMXBean().getInputArguments().asScala

      // not going to bother to check which one is last, - is not defined or + is defined libgraal will be active
      toCheck.exists(_ == "-XX:+UseJVMCINativeLibrary") || !toCheck.exists(
        _ == "-XX:-UseJVMCINativeLibrary"
      )
    }
  }
```
What you see is a bit of code in an abstract object initializer/constructor.
The `init()` can be overridden by a deriving class, the `system` has to be and will implicitly be used in the `init()` method.
The initialization of the used actor system, will try to obtain and access the object which is being created here.
If I understand the java memory model correctly I can safely do this, off thread, as it will ensure the object to be initalized before other threads can access it. Yet if I do this with libgraal active, it gets stuck if I don't insert a small wait. While it doesn't get in to that problem if I disable libgraal.

**Comment by mcraj017** [comment]
@spangaer Please provide the reproducer

**Comment by spangaer** [comment]
So, in the process of attempting to write a reproducer, I ended up figuring out it was a deadlock after-all.

**SHAME! SHAME! SHAME!**

I started an Akka `ActorSystem` whithin a scala `lazy val` field but got ambushed by a lazy val taking a global lock on the parent object, while `ActorSystem` creation blocked, waiting on logging init on another thread (which needs access to another lazy val in that object). 
I guess that the libgraal version just started the deadlocking thread faster then the jvmgraal version, the latter not running in to the same problem. 

My apologies for the disturbance.
